ai/recommender/surprise_recommenders.py

```diff
@@ -38,31 +44,122 @@ class SurpriseRecommender(BaseRecommender):
+                    + '_' + self.metric 
+                    + '_pmlb_20200505'
+                    +'.pkl.gz')
+
+        super().__init__(ml_type, metric, ml_p, filename=filename,
+                knowledgebase=knowledgebase, random_state=random_state)
+
+        # note: anything set after super().__init() will change the field 
+        # of any recommender from a file
+
         
-        self.first_fit = True
         self.min_epochs = 10
         self.max_epochs = 100
     
     @property
     def algo_name(self):
+        if type(self.algo).__name__ is None:
+            return type(self).__name__
         return type(self.algo).__name__
 
+    def set_training_data(self, results_data, results_mf=None, source='pennai'):
+        """Used for loading pickled recomenders to set results_df 
+        without training.
+        :param results_data: DataFrame with columns corresponding to:
+                'dataset'
+                'algorithm'
+                'parameters'
+                self.metric
+        :param results_mf: metafeatures for the datasets in results_data 
+        """
+        # update trained dataset models and hash table
+        super().update(results_data, results_mf, source)
+        self.update_training_data(results_data, shuffle=True)
+        rowHashes = hash_pandas_object(self.results_df).values 
+        newHash = hashlib.sha256(rowHashes).hexdigest()
+        if hasattr(self, 'results_df_hash'):
+            if newHash == self.results_df_hash:
+                logger.info('results_df hashes match')
+            else:
+                error_msg = 'the results_df hash from the pickle is different'
+                logger.error(error_msg)
+                raise ValueError(error_msg)
+            del self.results_df_hash
+
+    def load(self, filename=None, knowledgebase = None):
+        """Load a saved recommender state."""
+        if knowledgebase is None:
+            logger.warning('A knowledgebase needs to be provided to load '
+            'Surprise Recommenders from file. Not loading.')
+            return 
+        loaded = super().load(filename=filename)
+        if loaded:
+            logger.info('setting training data...')
+            self.set_training_data(knowledgebase, source='knowledgebase')
+
+    def save(self, filename=None):
+        """Save the current recommender."""
+
+        if filename is None:
+            fn = self.filename
+        else:
+            fn = filename
+        if os.path.isfile(fn):
+            logger.warning('overwriting ' + fn)
+
+        logger.info('saving recommender as ' + fn)
+        f = gzip.open(fn, 'wb')
+        save_dict = copy.deepcopy(self.__dict__)
+        # remove results_df to save space. this gets loaded by load() fn.
+        rowHashes = hash_pandas_object(save_dict['results_df']).values 
+        save_dict['results_df_hash'] = hashlib.sha256(rowHashes).hexdigest() 
+        print('save_dict[results_df]:',save_dict['results_df'].head())
```

Comment:
remove the `print` herein.

Also, since the hash tag checking is added to load() above. I think `results_df` should also be saved  for used it in load ().